src/github.com/getlantern/balancer/dialer.go

```diff
@@ -94,44 +70,83 @@ func (d *dialer) start() {
 					d.OnClose()
 				}
 				return
-			case t := <-d.errCh:
```

Comment:
I removed `errCh` and `markFailure` directly to avoid blocking caller, so goroutine is no longer required.